### PR TITLE
fix: concurrent map read-write on access module values

### PR DESCRIPTION
**Pull request by diafour**
#### Overview

Use getters with locks to get module values from ModuleManager.

#### What this PR does / why we need it

Fix panic on module values access:

```
fatal error: concurrent map read and map write

goroutine 19695 [running]:
runtime.throw(0x2b4ea10, 0x21)
        /usr/local/go/src/runtime/panic.go:1117 +0x72 fp=0xc007eea958 sp=0xc007eea928 pc=0x4432b2
runtime.mapaccess1_faststr(0x2626360, 0xc000877260, 0xc000cc22d7, 0x5, 0xc0117758c0)
        /usr/local/go/src/runtime/map_faststr.go:21 +0x465 fp=0xc007eea9c8 sp=0xc007eea958 pc=0x41ed65
github.com/flant/addon-operator/pkg/module_manager.(*Module).Values(0xc000112e60, 0xc000cc22d7, 0x5, 0x0)
        /go/pkg/mod/github.com/flant/addon-operator@v1.0.6-0.20220620123828-eaf343743c09/pkg/module_manager/module.go:741 +0x46b fp=0xc007eeab28 sp=0xc007eea9c8 pc=0x1a6b24b
github.com/flant/addon-operator/pkg/module_manager.(*ModuleHook).Run(0xc011887770, 0x2ad888a, 0x8, 0xc007423ce0, 0x1, 0x1, 0xc00eb4a240, 0xc013ef8030, 0xf10000c004781128, 0x1650933)
        /go/pkg/mod/github.com/flant/addon-operator@v1.0.6-0.20220620123828-eaf343743c09/pkg/module_manager/module_hook.go:261 +0xc1f fp=0xc007eeb0a8 sp=0xc007eeab28 pc=0x1a7161f
github.com/flant/addon-operator/pkg/module_manager.(*moduleManager).RunModuleHook(0xc000886f20, 0x35d831d, 0x34, 0x2ad888a, 0x8, 0xc007423ce0, 0x1, 0x1, 0xc00eb4a240, 0x0, ...)
        /go/pkg/mod/github.com/flant/addon-operator@v1.0.6-0.20220620123828-eaf343743c09/pkg/module_manager/module_manager.go:923 +0x445 fp=0xc007eeb138 sp=0xc007eeb0a8 pc=0x1a7ac45
github.com/flant/addon-operator/pkg/addon-operator.(*AddonOperator).HandleModuleHookRun(0xc00069e7e0, 0x2f364d8, 0xc0138e7d00, 0xc00eb4a240, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/pkg/mod/github.com/flant/addon-operator@v1.0.6-0.20220620123828-eaf343743c09/pkg/addon-operator/operator.go:1479 +0xa0e fp=0xc007eeba10 sp=0xc007eeb138 pc=0x1a9374e
github.com/flant/addon-operator/pkg/addon-operator.(*AddonOperator).TaskHandler(0xc00069e7e0, 0x2f364d8, 0xc0138e7d00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        
```

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Review by shvgn — COMMENTED** [review]
Awaiting for a couple of clarifications.

File: pkg/module_manager/module_manager.go [hunk 1]
```diff
@@ -71,8 +71,12 @@ type ModuleManager interface {
 	GlobalValuesPatches() []utils.ValuesPatch
 	UpdateGlobalConfigValues(configValues utils.Values)
 	UpdateGlobalDynamicValuesPatches(valuesPatch utils.ValuesPatch)
+
+	ModuleConfigValues(moduleName string) utils.Values
+	ModuleDynamicValuesPatches(moduleName string) []utils.ValuesPatch
 	UpdateModuleConfigValues(moduleName string, configValues utils.Values)
 	UpdateModuleDynamicValuesPatches(moduleName string, valuesPatch utils.ValuesPatch)
+	ApplyModuleDynamicValuesPatches(moduleName string, values utils.Values) (utils.Values, error)
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/module_manager/module_manager.go:79]: ApplyModuleDynamicValuesPatches is a new method. Previously, Values() method in module.go used internal structures from moduleManager. I've moved this code into ApplyModuleDynamicValuesPatches to add locking.

File: pkg/module_manager/module_manager.go [hunk 2]
```diff
@@ -1028,6 +1029,20 @@ func (mm *moduleManager) UpdateGlobalDynamicValuesPatches(valuesPatch utils.Valu
...
+	defer mm.valuesLayersLock.Unlock()
+	return mm.kubeModulesConfigValues[moduleName]
+}
+
+// ModuleConfigValues returns all patches for dynamic values.
+func (mm *moduleManager) ModuleDynamicValuesPatches(moduleName string) []utils.ValuesPatch {
+	mm.valuesLayersLock.Lock()
+	defer mm.valuesLayersLock.Unlock()
+	return mm.modulesDynamicValuesPatches[moduleName]
+}
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/module_manager/module_manager.go:1044]: `mm.modulesDynamicValuesPatches` was used directly from module.go. This PR replaces such usages with public methods.

File: pkg/module_manager/module_manager.go [hunk 3]
```diff
@@ -177,7 +181,7 @@ type moduleManager struct {
 	// Values from modules/values.yaml file
 	commonStaticValues utils.Values
 
-	// A lock to synchronize access to *ConfigValues and *DynamicValuesPatches fields.
+	// A lock to synchronize access to *ConfigValues and *DynamicValuesPatches maps.
 	valuesLayersLock sync.Mutex
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/module_manager/module_manager.go]: Good point! I've changed it to RWMutex in new commit.

File: pkg/module_manager/module_manager.go [hunk 4]
```diff
@@ -1045,6 +1058,24 @@ func (mm *moduleManager) UpdateModuleDynamicValuesPatches(moduleName string, val
 		valuesPatch)
 }
 
+func (mm *moduleManager) ApplyModuleDynamicValuesPatches(moduleName string, values utils.Values) (utils.Values, error) {
+	mm.valuesLayersLock.Lock()
+	defer mm.valuesLayersLock.Unlock()
```
Comment (yalosev) [inline pkg/module_manager/module_manager.go:1063]: do you really need write lock here?
Seems like you are reading patches map here and RLock is enough here

File: pkg/module_manager/module_manager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (shvgn) [reply pkg/module_manager/module_manager.go:79]: Hmm, this interface is used only once. I wonder if it is used anywhere else. And why it even exists :)

**Review by shvgn — APPROVED** [review]
LGTM, though the question by @yalosev should be addressed

File: pkg/module_manager/module_manager.go (hunk 4 above)
Comment (diafour) [reply pkg/module_manager/module_manager.go]: Changed to RLock.

**Pull request merged**